File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
def dfs(graph, node, res):
    while graph[node]:
        next_node = graph[node].pop()
        res.append(dfs(graph, next_node, res))
    
    return node
    

def solution(tickets):
    # 주어진 항공권을 모두 사용하여 경로를 탐색
    # 모든 도시를 방문할 수 없는 경우는 없음 -> 어떻게든 한줄로 이어진다는 뜻
    # graph를 형성하고 ICN부터 방문하여 dfs하면됨
    
    from collections import defaultdict
    
    graph = defaultdict(list)
    
    for a, b in tickets:
        graph[a].append(b)
    for k, _ in graph.items():
        graph[k].sort(reverse=True) # 알파벳 순으로 탐색하기 위함
    
    res = []

    dfs(graph, "ICN", res)
    
    return ["ICN"]+res[::-1]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;
import java.io.*;

class Solution {
    static String[][] tickets;
    static boolean[] visited;
    static Deque<String> res = new LinkedList<>();
    
    public static boolean dfs(String from){
        
        res.addLast(from);
        
        if (res.size()==tickets.length+1){
            return true;
        }
        
        for(int i=0; i<tickets.length; i++){
            if (tickets[i][0].equals(from) && !visited[i]){
                visited[i] = true;
                if (dfs(tickets[i][1])){
                    return true;
                }
                visited[i] = false;
            }
        }
        
        res.pollLast();
        
        return false;
    }
    
    
    public String[] solution(String[][] tickets) {
        this.tickets = tickets;
        visited = new boolean[tickets.length];
        
        Arrays.sort(this.tickets, (s1, s2) -> {
            if (s1[0].equals(s2[0])){
                return s1[1].compareTo(s2[1]);
            }
            return s1[0].compareTo(s2[0]);
        });
        
        dfs("ICN");
        
        String[] result = new String[tickets.length+1];
        
        for(int i=0; i<tickets.length+1;i++){
            result[i] = res.pollFirst();
        }
        
        return result;
    }
}
